File: app/src/main/java/app/cloudmusic/widget/PlayPauseView.java
package app.cloudmusic.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import app.cloudmusic.R;
import app.cloudmusic.utils.UIUtils;

/**
 * Created by Administrator on 2018/1/9.
 */

public class PlayPauseView extends View {
    public static final int PLAY_STATE_PLAYING = 1;
    public static final int PLAY_STATE_PAUSE = 0;

    private Paint paint;
    private int state = PLAY_STATE_PLAYING;//播放状态
    private Path path,path1,path2;//里面三角路径
    private int radius;//半径
    private int color = ContextCompat.getColor(getContext(), R.color.colorPrimary);
    private int max;//最大播放进度
    private int progress = 0;//播放进度

    public PlayPauseView(Context context) {
        this(context,null);
    }

    public PlayPauseView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        double hudu = 2*Math.PI/360;
        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStrokeWidth(5);
        paint.setStyle(Paint.Style.STROKE);
        radius = UIUtils.dip2px(getContext(),25)/2;
        path = new Path();
        path1 = new Path();
        path2 = new Path();
        int x = UIUtils.dip2px(getContext(),50)/2;
        int y = UIUtils.dip2px(getContext(),50)/2;
        int side = radius*2/3;
        int x1 = (int) (x - side/(2*Math.tan(hudu*60)));
        int y1 = y - side/2;
        path.moveTo(x1,y1);
        path1.moveTo(x1,y1);
        int x2 = x1;
        int y2 = y1 + side;
        path.lineTo(x2,y2);
        path1.lineTo(x2,y2);
        int x3 = (int) (x1+side*Math.sin(hudu*60));
        int y3 = y1+side/2;
        path.lineTo(x3,y3);
        int x4 = (int) (x1 + side*Math.tan(hudu*30));
        path2.moveTo(x4,y1);
        path2.lineTo(x4,y2);
        path.close();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(UIUtils.dip2px(getContext(),50),UIUtils.dip2px(getContext(),50));
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if(state == PLAY_STATE_PAUSE){
            paint.setColor(Color.parseColor("#444444"));
            canvas.drawCircle(canvas.getWidth()/2,canvas.getHeight()/2,radius,paint);
            canvas.drawPath(path,paint);
            paint.setColor(color);
            RectF rectF = new RectF(UIUtils.dip2px(getContext(),25)/2,UIUtils.dip2px(getContext(),25)/2,UIUtils.dip2px(getContext(),50)-UIUtils.dip2px(getContext(),25)/2,UIUtils.dip2px(getContext(),50)-UIUtils.dip2px(getContext(),25)/2);
            int angle = (int) (100*progress*3.6/max);
            canvas.drawArc(rectF,-90,angle,false,paint);
        }else{
            paint.setColor(Color.parseColor("#999999"));
            canvas.drawCircle(canvas.getWidth()/2,canvas.getHeight()/2,radius,paint);
            paint.setColor(color);
            canvas.drawPath(path1,paint);
            canvas.drawPath(path2,paint);
            RectF rectF = new RectF(UIUtils.dip2px(getContext(),25)/2,UIUtils.dip2px(getContext(),25)/2,UIUtils.dip2px(getContext(),50)-UIUtils.dip2px(getContext(),25)/2,UIUtils.dip2px(getContext(),50)-UIUtils.dip2px(getContext(),25)/2);
            int angle = (int) (100*progress*3.6/max);
            canvas.drawArc(rectF,-90,angle,false,paint);
        }
    }

    public void setMax(int max){
        this.max = max;
    }

    public void setProgress(int progress){
        this.progress = progress;
        postInvalidate();
    }

    public void setState(int state){
        this.state = state;
        postInvalidate();
    }
}

File: app/src/main/java/app/cloudmusic/widget/PlayPauseView.kt
package app.cloudmusic.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import androidx.core.content.ContextCompat
import android.util.AttributeSet
import android.view.View

import app.cloudmusic.R
import app.cloudmusic.utils.UIUtils

/**
 * Created by Administrator on 2018/1/9.
 */

class PlayPauseView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {

    private var paint: Paint? = null
    private var state = PLAY_STATE_PLAYING//播放状态
    private var path: Path? = null
    private var path1: Path? = null
    private var path2: Path? = null//里面三角路径
    private var radius: Int = 0//半径
    private val color = ContextCompat.getColor(getContext(), R.color.colorPrimary)
    private var max: Int = 0//最大播放进度
    private var progress = 0//播放进度

    init {
        init()
    }

    private fun init() {
        val hudu = 2 * Math.PI / 360
        paint = Paint(Paint.ANTI_ALIAS_FLAG)
        paint!!.strokeWidth = 5f
        paint!!.style = Paint.Style.STROKE
        radius = UIUtils.dip2px(context, 25f) / 2
        path = Path()
        path1 = Path()
        path2 = Path()
        val x = UIUtils.dip2px(context, 50f) / 2
        val y = UIUtils.dip2px(context, 50f) / 2
        val side = radius * 2 / 3
        val x1 = (x - side / (2 * Math.tan(hudu * 60))).toInt()
        val y1 = y - side / 2
        path!!.moveTo(x1.toFloat(), y1.toFloat())
        path1!!.moveTo(x1.toFloat(), y1.toFloat())
        val y2 = y1 + side
        path!!.lineTo(x1.toFloat(), y2.toFloat())
        path1!!.lineTo(x1.toFloat(), y2.toFloat())
        val x3 = (x1 + side * Math.sin(hudu * 60)).toInt()
        val y3 = y1 + side / 2
        path!!.lineTo(x3.toFloat(), y3.toFloat())
        val x4 = (x1 + side * Math.tan(hudu * 30)).toInt()
        path2!!.moveTo(x4.toFloat(), y1.toFloat())
        path2!!.lineTo(x4.toFloat(), y2.toFloat())
        path!!.close()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        setMeasuredDimension(UIUtils.dip2px(context, 50f), UIUtils.dip2px(context, 50f))
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        if (state == PLAY_STATE_PAUSE) {
            paint!!.color = Color.parseColor("#444444")
            canvas.drawCircle((canvas.width / 2).toFloat(), (canvas.height / 2).toFloat(), radius.toFloat(), paint!!)
            canvas.drawPath(path!!, paint!!)
            paint!!.color = color
            val rectF = RectF((UIUtils.dip2px(context, 25f) / 2).toFloat(), (UIUtils.dip2px(context, 25f) / 2).toFloat(), (UIUtils.dip2px(context, 50f) - UIUtils.dip2px(context, 25f) / 2).toFloat(), (UIUtils.dip2px(context, 50f) - UIUtils.dip2px(context, 25f) / 2).toFloat())
            val angle = (100.0 * progress.toDouble() * 3.6 / max).toInt()
            canvas.drawArc(rectF, -90f, angle.toFloat(), false, paint!!)
        } else {
            paint!!.color = Color.parseColor("#999999")
            canvas.drawCircle((canvas.width / 2).toFloat(), (canvas.height / 2).toFloat(), radius.toFloat(), paint!!)
            paint!!.color = color
            canvas.drawPath(path1!!, paint!!)
            canvas.drawPath(path2!!, paint!!)
            val rectF = RectF((UIUtils.dip2px(context, 25f) / 2).toFloat(), (UIUtils.dip2px(context, 25f) / 2).toFloat(), (UIUtils.dip2px(context, 50f) - UIUtils.dip2px(context, 25f) / 2).toFloat(), (UIUtils.dip2px(context, 50f) - UIUtils.dip2px(context, 25f) / 2).toFloat())
            val angle = (100.0 * progress.toDouble() * 3.6 / max).toInt()
            canvas.drawArc(rectF, -90f, angle.toFloat(), false, paint!!)
        }
    }

    fun setMax(max: Int) {
        this.max = max
    }

    fun setProgress(progress: Int) {
        this.progress = progress
        postInvalidate()
    }

    fun setState(state: Int) {
        this.state = state
        postInvalidate()
    }

    companion object {
        val PLAY_STATE_PLAYING = 1
        val PLAY_STATE_PAUSE = 0
    }
}
